lib/THC/THCTensorIndex.cu

```diff
@@ -94,6 +95,124 @@ __global__ void indexCopyLargeIndex(TensorInfo<T, IndexType> dst,
   }
 }
 
+template <typename T, size_t n>
+struct AtomicAddIntegerImpl;
+
+template<typename T>
+struct AtomicAddIntegerImpl<T, 1> {
+  __device__ void operator()(T *address, T val) {
+    unsigned int * address_as_ui =
+        (unsigned int *) (address - ((size_t)address & 3));
+    unsigned int old = *address_as_ui, sum, newval, assumed;
+    unsigned int selectors[] = {0x3214, 0x3240, 0x3410, 0x4210};
+    unsigned int sel = selectors[(size_t)address & 3];
+
+    do {
+      assumed = old;
+      sum = val +  (T)__byte_perm(old, 0, (size_t)address & 3 | 0x4440);
```

Comment:
what about overflow? doesn't seem correct to me since you could be bleeding into other bytes?